parsec/core/fs/storage/workspace_storage.py

```diff
@@ -64,15 +57,126 @@ def __init__(
         # Manifest and block storage
+        self.block_storage = block_storage
+        self.chunk_storage = chunk_storage
+
+    def _get_next_fd(self) -> FileDescriptor:
+        self.fd_counter += 1
+        return FileDescriptor(self.fd_counter)
+
+    async def get_manifest(self, entry_id: EntryID) -> LocalManifest:
+        raise NotImplementedError
+
+    # Locking helpers
+
+    @asynccontextmanager
+    async def lock_entry_id(self, entry_id: EntryID):
+        async with self.entry_locks[entry_id]:
+            try:
+                self.locking_tasks[entry_id] = hazmat.current_task()
+                yield entry_id
+            finally:
+                del self.locking_tasks[entry_id]
+
+    @asynccontextmanager
+    async def lock_manifest(self, entry_id: EntryID):
+        async with self.lock_entry_id(entry_id):
+            yield await self.get_manifest(entry_id)
+
+    def _check_lock_status(self, entry_id: EntryID) -> None:
+        task = self.locking_tasks.get(entry_id)
+        if task != hazmat.current_task():
+            raise RuntimeError(f"Entry `{entry_id}` modified without beeing locked")
+
+    # File management interface
+
+    def create_file_descriptor(self, manifest: LocalFileManifest) -> FileDescriptor:
+        assert isinstance(manifest, LocalFileManifest)
+        fd = self._get_next_fd()
+        self.open_fds[fd] = manifest.id
+        return fd
+
+    async def load_file_descriptor(self, fd: FileDescriptor) -> LocalFileManifest:
+        try:
+            entry_id = self.open_fds[fd]
+        except KeyError:
+            raise FSInvalidFileDescriptor(fd)
+        manifest = await self.get_manifest(entry_id)
+        assert isinstance(manifest, LocalFileManifest)
+        return manifest
+
+    def remove_file_descriptor(self, fd: FileDescriptor) -> None:
+        try:
+            self.open_fds.pop(fd)
+        except KeyError:
+            raise FSInvalidFileDescriptor(fd)
+
+    # Block interface
+
+    async def set_clean_block(self, block_id: BlockID, block: bytes) -> None:
+        assert isinstance(block_id, BlockID)
+        return await self.block_storage.set_chunk(ChunkID(block_id), block)
+
+    async def clear_clean_block(self, block_id: BlockID) -> None:
+        assert isinstance(block_id, BlockID)
+        try:
+            await self.block_storage.clear_chunk(ChunkID(block_id))
+        except FSLocalMissError:
+            pass
+
+    async def get_dirty_block(self, block_id: BlockID) -> bytes:
+        return await self.chunk_storage.get_chunk(ChunkID(block_id))
+
+    # Chunk interface
+
+    async def get_chunk(self, chunk_id: ChunkID) -> bytes:
+        assert isinstance(chunk_id, ChunkID)
+        try:
+            return await self.chunk_storage.get_chunk(chunk_id)
+        except FSLocalMissError:
+            return await self.block_storage.get_chunk(chunk_id)
+
+    async def set_chunk(self, chunk_id: ChunkID, block: bytes) -> None:
+        assert isinstance(chunk_id, ChunkID)
```

Comment:
assert still needed ?

Reply:
Maybe non but mypy does not cover the entire code base, so it might be useful to keep it.